readme.txt
Compiler-Generator Coco/R
=========================

               (Pat Terry,  updated Sat  03-11-00)
                       p.terry@ru.ac.za

Coco/R generates recursive descent parsers and their associated scanners from
attributed grammars.  The distribution kit contains source files for the
compiler generator Coco/R (Moessenboeck, 1990), as well as the executables,
frame files, and some sample applications.

Coco/R was originally written in Oberon by Hanspeter Moessenboeck.  It was
then ported into Modula-2; the first such port being done by Hanspeter
Moessenboeck for the Apple MacMeth system.  A port was done to JPI TopSpeed
Modula-2 at ETH Zurich by Marc Brandis and Christof Brass.  This was made
available to Pat Terry, who corrected several errors, enhanced the
portability, and provided the versions which you find on this distribution.
Responsibility for remaining bugs in these versions now rests largely with
Pat Terry.

The Modula-2 versions for WinTel platforms are known to be immediately
usable with any of the PIM based TopSpeed (JPI) compilers; the Fitted
Software Tools (FST) compilers, versions 2.x through 3.5; the StonyBrook
QuickMod 2.2 compiler; and the Logitech compiler, version 3.03.  As from
release 1.43, the system should also be fully compatible with ISO compilers
such as StonyBrook V4, and XDS from XTech Ltd.

Full source code for Coco/R compatible with these compilers is supplied.  It
is hoped that this will be trivially portable to other Modula-2 compilers
presently available, save for the I/O module used by Coco/R (and, in one
mode of use, its generated parsers); code for the I/O module for the
compilers listed above is, naturally, supplied.

So far as other operating systems are concerned, the Modula-2 versions of
Coco/R have also been ported to Mocka Modula-2 by Pat Terry and Toshinori
Maeno, for both Linux and BSD386, and to the Apple MAC p1 compiler by Woody
Yeung (yeung@reed.edu).  The source code is identical save for the I/O
modules for these compilers (which code is also provided in the DOS kit),
and in addition, distributions are available for these systems in their
native format.

Coco/R was first ported to Gardens Point Modula-2 by John Gough. (Gardens
Point Modula-2 is available for a wide variety of platforms, including Intel,
Sparc and MIPS machines).  The source code for the I/O modules for use with
the freeware Gardens Point Modula-2 for the PC (gpm-pc) is supplied with the
DOS distribution kit, and a complete port of Coco/R for the Unix hosted
versions of Gardens Point Modula-2 is also available.

Coco/R was also ported to TDI Modula-2 for the Atari by Rolf Schrader.
The sources (at level 1.39) are also supplied in a complete kit.  These
have numerous minor differences from the MS-DOS versions, although they
were derived from them, and not having access to an Atari, they have not
been kept up to date.  A contact address for this version is

    Rolf Schrader, Dipl.-Math. & Dipl.-Phys.,
    Kronenstr. 24,
    D-76133 Karlsruhe,
    Germany.

As from release 1.39, versions of Coco/R that produce Turbo Pascal units have
been made available through a port largely done by Volker Pohlers
(pohlers@escher.hrz.fh-stralsund.de), and now also maintained by Pat Terry.

There is also a C/C++ version of Coco/R, ported by Frankie Arzu
(frankie_arzu@yahoo.com or farzu@uvg.edu.gt).  As from version 1.06 a user has
had the option of generating either C or C++ code. The C++ version generates
scanner and parser classes based on a simple but effective class hierarchy.

All of these versions of Coco/R can bootstrap themselves to generate a
driver, parser, scanner, and semantic evaluator from an attributed grammar
CR.ATG.  This grammar thus serves as an an example of how to write compiler
descriptions for Coco/R.  There are also other simpler examples showing its
use.

Oberon versions of Coco/R are still available, and a new Java version has
recently been released.  These original versions are maintained by Hanspeter
Moessenboeck in Linz (see address below); an extended Java one has been
derived by Pat Terry.

It is important to realise that Coco/R is intended for use with grammars that
meet the LL(1) conditions.  Many grammars require some massaging before these
conditions are met.  Some of the examples in the kit are (deliberately)
non-LL(1), and are intended as examples for study and experiment.

While every attempt has been made to ensure that Coco/R performs
satisfactorily, the developers can accept no liability for any damage or
loss, including special, incidental, or consequential, caused by the use of
the software, directly or indirectly.


Where to get the latest versions
================================

Up to date information about Coco/r is available on the WWW at

        http://cs.ru.ac.za/homes/cspt/cocor.htm

The latest versions of Coco/R should be available for anonymous ftp from any
of the servers listed below (not all servers carry all versions):

In Europe

         ftp.ssw.uni-linz.ac.at:/pub/Coco

In Australia

         ftp.fit.qut.edu.au:/pub/coco

The freeware gpm-pc Modula-2 compiler is available on the same server.

In South Africa

         cs.ru.ac.za:/pub/coco

The latest shareware version of the FST Modula-2 compiler for the PC is
available on the same server in the directory /pub/languages.

In Central America

         uvg.edu.gt:/pub/coco

In the USA

         ftp.psg.com:/pub/modula-2/coco

where it also will be available by mail server

         To: server@ftp.psg.com
         From: <address to which it is to be sent>
         Subject: send pub/modula-2/coco/...

The latest shareware version of FST Modula-2 for the PC is available on the
same server in the directory /pub/modula-2 (fst-40s.lzh, compressed with
LHA).

The distributions come in various files (in each case "xxx" denotes the
release number, for example 143 for version 1.43)

    (a)    COCOxxx.EXE   - the kit for MS-DOS platforms
    (b)    gpmxxx.tgz    - the equivalent kit for Gardens Point
    (c)    ATARIxxx.EXE  - the kit for the Atari platform
    (d)    MOCKAxxx.tgz  - the kit for Mocka (Linux and 386BSD)
    (e)    COCOPxxx.EXE  - the kit for producing Turbo Pascal hosted in Modula
    (f)    TURBOxxx.EXE  - the kit for producing Turbo Pascal hosted in Pascal
    (g)    COCORCxx.ZIP  - the kit for producing C/C++
    (g)    README        - this file
    (h)    README.1st    - unpacking and installation notes

The kits contain sources, objects, examples and vanilla-ASCII documentation
files.  The MS-DOS versions are self-extracting files for MS-DOS, compressed
with the public domain utility LHA (which is also available on many Unix
systems).  All that is needed to install the system on a WinTel platform
is to copy the .EXE file to a suitable directory and to execute it.  About
1.6 MB of space is needed for the entire distribution, but several files
are compiler/platform specific, and need not be retained.

The Atari version is also supplied as a self-extracting file for MS-DOS.  It
is hoped that users will be able to find an easy way to transfer the files to
their Atari systems.  The Atari port is presently sufficiently different from
the others that the distribution kit is self contained, rather than being
merged with the MS-DOS versions.

For the latest on the Oberon versions, point your browser at

         ftp://ftp.ssw-linz.ac.at/pub/Oberon/LinzTools/Coco.cod

or look for the ile on the other mirrors.  Like the Modula-2 version, this
release comes with full source code.

For the Java versions, point your browser at

         http://www.ssw-linz.ac.at/Projects/Coco/Coco.html

or

         http://cs.ru.ac.za/homes/cspt/javacoco.htm

for the extended version.

FileList
========

The following gives an indication of the files provided in various kits of
Coco/R (Modula).  Essentially the same distribution is to be found in the
other kits (including the Pascal versions) - allowing for differences in
file and directory naming conventions.  There may also be extra files,
depending on the release date and version.  The Pascal kits have "units"
with extensions PAS, rather than separate DEF and MOD files.

  READ.ME              This file
  README.GPM           Notes on the Gardens Point implementation
  README.ISO           Notes on using the ISO compliant system
  README.1ST           Unpacking and installation information
  HELP.LHA             Notes on packing the kit with LHA
  PACK.BAT             Pack up system into a self-extracting file
  COCOR.EXE            Compiler-compiler executable
  HISTORY              Summary of bug fixes and modifications
  CONVERT.MOD          Program to convert text files between systems -
                       translating line marks appropriate to the O/S in use

Subdirectory "Docs"    Documentation

  COCO.USE             Notes on the use of Coco
  COCOL                The compiler description language
  CHANGES              Differences between Oberon and Modula-2 versions

Subdirectories "Frames"  and "IsoFrame"  Specimen frame files.  One set is
for generating parsers to be used with the supplied FileIO library, the
other set is for generating parsers to be used with an ISO I/O library.

  SCANNER.FRM          Frame from which the standard scanner is generated
  SCANNERC.FRM         Frame from which a scanner can be generated
                       allowing the retrieval of comment text
  SCANNER2.FRM         Frame from which a more efficient scanner can be
                       generated for short source files
  PARSER.FRM           Frame from which the parser is generated
  COMPILER.FRM         Frame from which a driver compiler may be generated
                       (tabs in the listing files are retained)
  COMPILE2.FRM         Frame from which a driver compiler may be generated
                       (tabs in the listing files are expanded)

Subdirectory "Sources" and its subdirectories - Sources for Coco/R

  CR.MOD               Main program
  CR.ATG               Fully attributed grammar of Coco/R
  CR0.ATG              Simplifed grammar of Coco/R (no attributes)
  CRE.TXT              Error numbers and corresponding error message texts
  CRS.DEF + MOD        Scanner (generated by Coco/R from CR.ATG)
  CRP.DEF + MOD        Parser (generated by Coco/R from CR.ATG)
                       (The JPI specific version of CRP.MOD is in the
                       Sources\JPI directory)
  CRT.DEF + MOD        Symbol table, top-down graph, grammar tests.
  CRA.DEF + MOD        Generation of the scanner.
  CRX.DEF + MOD        Generation of the parser.
  CRC.DEF + MOD        Generation of the compiler driver.
  CRG.DEF + MOD        Module giving symbolic names for scanner/parser
                       to use (Generated with $N pragma)
  FileIO.DEF           General input output library definition
  FileIO.MOD           Source for FileIO.MOD appears in compiler specific
                       subdirectories - Sources\JPI Sources\FST Sources\STO
                       Sources\LOG or Sources\GPM etc
  Sets.DEF + MOD       Various set operations
  CR.FRM               Frame from which CR.MOD may be regenerated ($C pragma)

Subdirectory "Taste"   A tiny compiler and utilities as examples of Coco/R

  Taste.ATG            Compiler description
  Taste.FRM            Frame from which TASTE.MOD may be generated ($C pragma)
  TC.DEF + MOD         Code generator/interpreter
  TL.DEF + MOD         Symbol table handler

  Pretty.ATG           Pretty printer description
  Pretty.FRM           Frame from which PRETTY.MOD may be generated ($C pragma)

history.txt
Change History since version 1.21
=================================

Version 1.22 - internal, not released

Version 1.23 - Sun  05-31-1992

-   Updated and corrected documentation.
-   Scanner now returns "sym" of CARDINAL not INTEGER type.
-   Frame files modified to handle this change.
-   CRP.MOD and CRS.DEF and CRS.MOD modified to handle this change.
-   CRX.GenCond modified to handle small sets more safely.
-   CRS.NextCh in scanner now translates CHR(26) to CHR(0) so that DOS
    editors that append CHR(26) to text files will not cause problems.
-   Modified CHARACTERS definition on CR.ATG to allow only printable
    characters in strings (in particularly preventing strings from extending
    over more than one line).
-   Corrected errors in FileIO.WriteText in all four versions.
-   Corrected bug in CRA.ImportSymConsts that was generating too long a name
    for the generated Symbolic constants file.
-   Improved CRL.PrintErr to remove redundant parameter.
-   Fixed CRL.GetLine to ensure that CHR(26) was treated properly.

Version 1.24 - Mon  06-15-1992

-  Changed the $C pragma to $N (Names).  C had no mnemonic content!
-  Updated and corrected documentation.
-  Modified parser generation to avoid exporting symSetSize from the
   definition module.
-  Modified scanner yet again to convert CHR(26) to CHR(0) whenever found
   (thus avoiding problems caused by DOS editors that add CTRL-Z to the ends
   of text files more systematically).
-  Added pragma/command line option to allow suppression of definition module
   generation (recreating these each time means all clients have to be
   recompiled every time, which is tedious).
-  Added pragma/command line option to allow optional creation of a complete
   driver program (using frame file COMPILER.FRM) which incorporates a list
   handler with interleaved error messages.  A simple COMPILER.FRM is supplied
   that is "generic", but needs to be adapted for most applications.  Special
   tuned cases of COMPILER.FRM are provided for the sample grammars in the
   kit.
-  Modified the output directed to file <Grammar>E.TXT slightly.
-  Rearranged CRT.DEF and CRT.MOD to eliminate a superfluous procedure
   (PrintSet no longer exported, PrintSetTerminal eliminated).
-  Rearranged order of generation of def and imp mods for scanner and parser,
   and made minor changes to SCANNER.FRM and PARSER.FRM.
-  Moved export of "directory" and "lst" to scanner module CRS.DEF, and
   eliminated CRL.DEF and CRL.MOD.  The list handling is now handled in
   CR.MOD.
-  Various minor changes to CR.MOD made to reflect all these changes in the
   help screen, etc.

Version 1.25 - Sat  06-20-1992

- Minor bug in COMPILER.FRM files repaired

Version 1.26 - Wed  08-26-1992

- Rearranged COMPILER.FRM files to incorporate error/list handler as
  internal module.
- Replaced curious WriteText call in error/list handler with WriteString.
- Repaired bug in CRT.MOD to handle generation of names for terminals with
  single quotes   (Thanks to Tim Channon for pointing this one out!).

Version 1.27 - Tue  10-06-1992

- Extended length of name field in CRT.DEF to 40 chars.
- resultant small changes in output format in CRT.
- fixed bug in symbolic name generation for terminals with non-letters in
  them.
- suppressed generation of GrammarG.DEF along with other DEF mods when
  suppression requested.
- allowed COMPILER.FRM to default first to GRAMMAR.FRM in grammar directory
  (Thanks to Tim Channon for suggestion).
- fixed silly error in SCANNER.FRM that cause errors on first line to be
  reported in the wrong places.
- improved COMPILER.FRM files to treat tabs as spaces in error listing and
  thereby line up error messages more accurately (still needs improving).

Version 1.28 - Fri  04-30-1993

- Fixed bug in Term that resulted in program looping if alternatives in
  productions also involved unrecognisable tokens (Thanks to John Gough
  for discovering the bug).

Version 1.29 - Tue  07-13-1993

- Fixed various places where BITSET was omitted in front of default {} set
  constructors - this will be needed in some implementations.
- Changed WriteInt(f,x,0) calls to WriteInt(f,x,1) calls in various places
  (New standard requires ,0 to generate exactly one leading space).

Version 1.30 - Sat  08-07-1993

- Fixed bug that regarded two terminals like 'hello' and "hello" to be
  regarded as distinct (Thanks to Stuart van Onselen for spotting this one).
  Terminals typed as 'singlequoted' are now converted internally to
  "doublequoted" (except, obviously, where a double quote appears within
  'single"quoted"strings').

Version 1.31 - Sat  08-21-1993

- Various nasties.  Thanks to John Gough who pointed out an abuse of CRT.Set
  in CRA.DeleteRedundantStates.  In chasing this and testing on grammars that
  yield large numbers of states, various other problems came to light.  Moved
  the symbol table to the heap (CRT) and extended various limitations.
  CRA.GetstateSet was rewritten to conform to the Oberon Sources I have, and
  this seems to clear a nasty problem in melting states.  A lurking bug in
  MakeUnique has not yet been fully tracked.  Compiler restriction messages
  and CRX.WriteStatistics improved and extended.  Thus CRA, CRX and CRT have
  all been changed.

  Considering change to FileIO to replace "con" with StdIn and StdOut.

Version 1.32 - Sat  09-04-1993

- Major overhaul of CRA and CRS following debugging exercise by Hanspeter
  Moessenboeck in Zurich.  Hopefully this has cleared the automaton problems.
  His Oberon changes were adapted for the Modula-2 version, needing
  slight changes.  Also fixed potential problem in SCANNER.FRM to prevent
  garbage being produced on attempts to read past the end of the source buffer.

Version 1.33 - Tue  12-14-1993


src/CRQ.pas
(*$M 32000, 0, 655360 *)  (* Turbo Pascal directives *)

(* CR   Main Module of Coco/R
   ==   =====================

   This is a compiler generator that produces a scanner and a parser
   from an attributed grammar, and optionally a complete small compiler.

   Original code in Oberon by Hanspeter Moessenboeck, ETH Zurich
   Ported at ETH to Apple Modula, and thence to JPI-2 Modula.

   JPI version of 27 January 1991 was then modified to make more
   portable by Pat Terry, January - October 1992

   This is the IBM-PC MS-DOS Turbo Pascal version for generating Turbo Pascal
   based on the port first done by Volker Pohlers, October 1995

   This version outputs error messages in "standard" form for use with editors

Usage:
          COCOR [-options] GrammarName[.atg] [$options]

Input:
  attributed grammar   input grammar
  scanner.frm          frame file
  parser.frm           frame file
  compiler.frm         frame file (optional)

(Frame files must be in the sme directory as the grammar, or may be
found on a path specified by DOS environment variable CRFRAMES).

Output:
  <GrammarName>S.pas  generated scanner
  <GrammarName>P.pas  generated parser
  <GrammarName>.err   error numbers and corresponding error messages
  <GrammarName>.lst   source listing with error messages and trace output

Optionally

  <GrammarName>G.pas  generated symbolic names
  <GrammarName>.pas   generated compiler main module

Implementation restrictions
  1  too many nodes in graph (>1500)                 CRTable.NewNode
  2  too many symbols (>500)                         CRTable.NewSym, MovePragmas
  3  too many sets (>256 ANY-syms or SYNC syms)      CRTable.NewSet,
  4  too many character classes (>250)               CRTable.NewClass
  5  too many conditions in generated code (>100)    CRX.NewCondSet
  6  too many token names in "NAMES" (>100)          CRTable.NewName
  7  too many states in automata (>500)              CRA.NewState

Trace output
(To activate a trace switch, write "${letter}" in the input grammar, or
invoke Coco with a second command line parameter)

  A  Prints states of automaton

  C  Generates complete compiler module

  F  Prints start symbols and followers of nonterminals.

  G  Prints the top-down graph.

  I  Trace of start symbol set computation.

  L  Forces a listing (otherwise a listing is only printed if errors are found).
  
  N  Uses default names for symbol value constants.  This generates an
     extra module <grammar name>G, and corresponding import statements
     using constant names instead of numbers for symbols in parser and
     scanner.
     The constants are used unqualified and hence all needed constants
     have to be imported; so a complete import list for these constants
     is generated.
     There is no decision whether a constant is actually needed.

     The default conventions are (only terminals or pragmas can have names):
     single character   -->  <ASCII name (lowercase)>Sym
          eg. "+"       -->  plusSym
     character string   -->  <string>Sym
          eg. "PROGRAM" -->  PROGRAMSym
     scanner token      -->  <token name>Sym
          eg. ident     -->  identSym

  O  Trace of follow set computation (not yet implemented).

  P  Generates parser only

  Q  Generates error messages in quiet form for interfacing with editor

  S  Prints the symbol list.

  T  Suppresses generation of units (grammar tests only).

  X  Prints a cross reference list.

==========================================================================*)

PROGRAM CR;

USES CRS, (* lst, src, errors, directory, Error, CharAt *)
     CRP, (* Parse *)
     CRC, CRTable, CRA, CRX, FileIO;

CONST
  ATGExt = '.atg';
  LSTExt = '.lst';
  Version = '1.53q (for Pascal)';
  ReleaseDate = '17 September 2002';

VAR
  Options, GrammarName, ATGFileName, lstFileName : STRING;
  ll1 : BOOLEAN; (* TRUE, if grammar is LL(1) *)
  IDE, ok : BOOLEAN;  (* TRUE, if grammar tests ok so far *)
  P : INTEGER;   (* ParamCount *)

(* ------------------- Source Listing and Error handler -------------- *)

  TYPE
    CHARSET = SET OF CHAR;
    Err = ^ErrDesc;
    ErrDesc = RECORD
      nr, line, col: INTEGER;
      next: Err
    END;

  CONST
    TAB  = #09;
    _LF  = #10;
    _CR  = #13;
    _EF  = #0;
    LineEnds : CHARSET = [_CR, _LF, _EF];

  VAR
    firstErr, lastErr: Err;
    Extra : INTEGER;

  PROCEDURE StoreError (nr, line, col: INTEGER; pos: LONGINT); FAR;
  (* Store an error message for later printing *)
    VAR
      nextErr: Err;
    BEGIN
      NEW(nextErr);
      nextErr^.nr := nr; nextErr^.line := line; nextErr^.col := col;
      nextErr^.next := NIL;
      IF firstErr = NIL
        THEN firstErr := nextErr
        ELSE lastErr^.next := nextErr;
      lastErr := nextErr;
      INC(errors)
    END;

  PROCEDURE GetLine (VAR pos  : LONGINT;
                     VAR line : STRING;
                     VAR eof  : BOOLEAN);
  (* Read a source line. Return empty line if eof *)
    VAR
      ch: CHAR;
      i:  INTEGER;
    BEGIN
      i := 1; eof := FALSE; ch := CharAt(pos); INC(pos);
      WHILE NOT (ch IN LineEnds) DO BEGIN
        line[i] := ch; INC(i); ch := CharAt(pos); INC(pos)
      END;
      line[0] := Chr(i-1);
      eof := (i = 1) AND (ch = _EF);
      IF ch = _CR THEN BEGIN (* check for MsDos *)
        ch := CharAt(pos);
        IF ch = _LF THEN BEGIN INC(pos); Extra := 0 END
      END
    END;

  PROCEDURE PrintErr (line : STRING; nr, col: INTEGER);
  (* Print an error message *)

    PROCEDURE Msg (s: STRING);
      BEGIN
        Write(lst, s)
      END;

    PROCEDURE Pointer;
      VAR
        i : INTEGER;
      BEGIN
        Write(lst, '*****  ');
        i := 0;
        WHILE i < col + Extra - 2 DO BEGIN
          IF line[i] = TAB
            THEN Write(lst, TAB)
            ELSE Write(lst, ' ');
          INC(i)
        END;
        Write(lst, '^ ')
      END;

    BEGIN
      IF NOT IDE THEN Pointer;
      CASE nr OF
          0 : Msg('EOF expected');
          1 : Msg('ident expected');
          2 : Msg('string expected');
          3 : Msg('badstring expected');
          4 : Msg('number expected');
          5 : Msg('"COMPILER" expected');
          6 : Msg('"USES" expected');
          7 : Msg('"," expected');
          8 : Msg('";" expected');
          9 : Msg('"PRODUCTIONS" expected');
         10 : Msg('"=" expected');
         11 : Msg('"." expected');
         12 : Msg('"END" expected');
         13 : Msg('"CHARACTERS" expected');
         14 : Msg('"TOKENS" expected');
         15 : Msg('"NAMES" expected');
         16 : Msg('"PRAGMAS" expected');
         17 : Msg('"COMMENTS" expected');
         18 : Msg('"FROM" expected');
         19 : Msg('"TO" expected');
         20 : Msg('"NESTED" expected');
         21 : Msg('"IGNORE" expected');
         22 : Msg('"CASE" expected');
         23 : Msg('"+" expected');
         24 : Msg('"-" expected');
         25 : Msg('".." expected');
         26 : Msg('"ANY" expected');
         27 : Msg('"CHR" expected');
         28 : Msg('"(" expected');
         29 : Msg('")" expected');
         30 : Msg('"|" expected');
         31 : Msg('"WEAK" expected');
         32 : Msg('"[" expected');
         33 : Msg('"]" expected');
         34 : Msg('"{" expected');
         35 : Msg('"}" expected');
         36 : Msg('"SYNC" expected');
         37 : Msg('"CONTEXT" expected');
         38 : Msg('"<" expected');
         39 : Msg('">" expected');
         40 : Msg('"<." expected');
         41 : Msg('".>" expected');
         42 : Msg('"(." expected');
         43 : Msg('".)" expected');
         44 : Msg('not expected');
         45 : Msg('invalid TokenFactor');
         46 : Msg('invalid Factor');
         47 : Msg('invalid Factor');
         48 : Msg('invalid Term');
         49 : Msg('invalid Symbol');
         50 : Msg('invalid SingleChar');
         51 : Msg('invalid SimSet');
         52 : Msg('invalid NameDecl');
         53 : Msg('this symbol not expected in TokenDecl');
         54 : Msg('invalid TokenDecl');
         55 : Msg('invalid Attribs');
         56 : Msg('invalid Declaration');
         57 : Msg('invalid Declaration');
         58 : Msg('invalid Declaration');
         59 : Msg('this symbol not expected in CR');
         60 : Msg('invalid CR');
       
        101 : Msg('character set may not be empty');
        102 : Msg('string literal may not extend over line end');
        103 : Msg('a literal must not have attributes');
        104 : Msg('this symbol kind not allowed in production');
        105 : Msg('attribute mismatch between declaration and use');
        106 : Msg('undefined string in production');
        107 : Msg('name declared twice');
        108 : Msg('this type not allowed on left side of production');
        109 : Msg('earlier semantic action was not terminated');
        111 : Msg('missing production for grammar name');
        112 : Msg('grammar symbol must not have attributes');
        113 : Msg('a literal must not be declared with a structure');
        114 : Msg('semantic action not allowed here');
        115 : Msg('undefined name');
        116 : Msg('attributes not allowed in token declaration');
        117 : Msg('name does not match grammar name');
        118 : Msg('unacceptable constant value');
        119 : Msg('may not ignore CHR(0)');
        120 : Msg('token might be empty');
        121 : Msg('token must not start with an iteration');
        122 : Msg('comment delimiters may not be structured');
        123 : Msg('only terminals may be weak');
        124 : Msg('literal tokens may not contain white space');
        125 : Msg('comment delimiter must be 1 or 2 characters long');
        126 : Msg('character set contains more than one character');
        127 : Msg('could not make deterministic automaton');
        128 : Msg('semantic action text too long - please split it');
        129 : Msg('literal tokens may not be empty');
        130 : Msg('IGNORE CASE must appear earlier');
        ELSE BEGIN Msg('Error: '); Write(lst, nr) END
      END;
      WriteLn(lst);
    END;

  PROCEDURE PrintListing;
  (* Print a source listing with error messages *)
    VAR
      nextErr:   Err;
      eof:       BOOLEAN;
      lnr, errC: INTEGER;
      srcPos:    LONGINT;
      line:      STRING;
    BEGIN
      IF NOT IDE THEN BEGIN WriteLn(lst, 'Listing:'); WriteLn(lst) END;
      srcPos := 0; nextErr := firstErr;
      GetLine(srcPos, line, eof); lnr := 1; errC := 0;
      WHILE NOT eof DO BEGIN
        IF NOT IDE THEN WriteLn(lst, lnr:5, '  ', line);
        WHILE (nextErr <> NIL) AND (nextErr^.line = lnr) DO BEGIN
          IF IDE THEN
            Write(lst, ATGFileName, ' (', lnr:1, ', ', nextErr^.col - 1, ') ');
          PrintErr(line, nextErr^.nr, nextErr^.col); INC(errC);
          nextErr := nextErr^.next
        END;
        GetLine(srcPos, line, eof); INC(lnr);
      END;
      IF nextErr <> NIL THEN BEGIN
        IF NOT IDE THEN WriteLn(lst, lnr:5);
        WHILE nextErr <> NIL DO BEGIN
          IF IDE THEN
            Write(lst, ATGFileName, ' (', lnr:1, ', ', nextErr^.col - 1, ') ');
          PrintErr(line, nextErr^.nr, nextErr^.col); INC(errC);
          nextErr := nextErr^.next
        END
      END;
      IF NOT IDE AND (errC > 0) THEN
        BEGIN
          WriteLn(lst);
          Write(lst, errC:5, ' error');
          IF errC <> 1 THEN Write(lst, 's');
          WriteLn(lst); WriteLn(lst); WriteLn(lst);
        END;
    END;

PROCEDURE SetOption (s : STRING);
(* Set compiler options *) 
  VAR
    i : INTEGER;
  BEGIN
    FOR i := 2 TO Length(s) DO
      BEGIN
        s[i] := UpCase(s[i]);
        IF s[i] IN ['A' .. 'Z'] THEN CRTable.ddt[s[i]] := TRUE
      END;
  END;

PROCEDURE Msg (S : STRING);
  BEGIN
    WriteLn(S);
  END;

(* --------------------------- Help ------------------------------- *)

PROCEDURE Help;
  BEGIN
    WriteLn('Coco/R (MS-DOS) - Compiler-Compiler V', Version);
    WriteLn('Turbo Pascal (TM) version by Pat Terry/Volker Pohlers ', ReleaseDate);
    Msg('Usage: COCOR [-Options] [Grammar[.atg]] [-Options]');
    Msg('Example: COCOR -cs Test');
    Msg('');
    Msg('Options are');
    Msg('a  - Trace automaton');
    Msg('c  - Generate compiler module');
    Msg('f  - Give Start and Follower sets');
    Msg('g  - Print top-down graph');
    Msg('i  - Trace start set computations');
    Msg('l  - Force listing');
    Msg('n  - Generate symbolic names');
    Msg('p  - Generate parser only');
    Msg('q  - Generate error messages to interface with editor');
    Msg('s  - Print symbol table');
    Msg('t  - Grammar tests only - no code generated');
    Msg('x  - Print cross reference list');
    Msg('COMPILER.FRM, SCANNER.FRM and PARSER.FRM must be in the working directory,');
    Msg('or on the path specified by the environment variable CRFRAMES');
  END;

BEGIN
  firstErr := NIL; Extra := 1; P := 1;
  GrammarName := ParamStr(1);
  IF (GrammarName = '?') OR (GrammarName = '') OR (GrammarName = '/?') THEN
    BEGIN Help; HALT END;
  WHILE (Length(GrammarName) > 0) AND (GrammarName[1] IN ['-', '/']) DO BEGIN
    (* accept options before filename *)
    SetOption(GrammarName);
    INC(P); GrammarName := ParamStr(P);
  END;
  ok := GrammarName <> '';
  REPEAT
    IF NOT ok THEN
      BEGIN
        Write('Grammar[.atg] ? : ');
        ReadLn(GrammarName);
        IF GrammarName = '' THEN HALT;
      END;
    FileIO.AppendExtension(GrammarName, ATGExt, ATGFileName);
    GrammarName := ATGFileName;
    Assign(src, GrammarName);
    {$I-} Reset(src, 1); {$I+}
    ok := IOResult = 0;
    IF NOT ok THEN WriteLn('File <', GrammarName, '> not found.');
  UNTIL ok;
  INC(P); Options := ParamStr(P);
  IF Options <> '' THEN SetOption(Options);
  IDE := CRTable.ddt['Q'];
  FileIO.ExtractDirectory(GrammarName, directory);
  IF IDE
    THEN lstFileName := ''
    ELSE FileIO.ChangeExtension(GrammarName, LSTExt, lstFileName);
  FileIO.Open(lst, lstFileName, TRUE);
  WriteLn(lst, 'Coco/R - Compiler-Compiler V', Version);
  IF NOT IDE THEN
    BEGIN
      WriteLn(lst, 'Turbo Pascal (TM) version by Volker Pohlers/Pat Terry ', ReleaseDate);
      WriteLn(lst, 'Source file: ', GrammarName);
      WriteLn(lst);
      WriteLn; WriteLn('parsing file ', GrammarName)
    END;
  CRS.Error := StoreError;
  CRP.Parse;
  IF errors = 0
    THEN
      BEGIN
        IF NOT IDE
          THEN
            BEGIN
              Msg('testing grammar');
              WriteLn(lst, 'Grammar Tests:'); WriteLn(lst);
            END
          ELSE
            BEGIN
              WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) Grammar tests');
            END;
        CRTable.CompSymbolSets;
        IF IDE THEN
          BEGIN
            WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) Undefined tests');
          END;
        CRTable.TestCompleteness(ok);
        IF ok THEN
          BEGIN
            IF IDE THEN
              BEGIN
                WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) Unreachable tests');
              END;
            CRTable.TestIfAllNtReached(ok)
          END;
        IF ok THEN
          BEGIN
            IF IDE THEN
              BEGIN
                WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) Circular tests');
              END;
            CRTable.FindCircularProductions(ok)
          END;
        IF ok THEN
          BEGIN
            IF IDE THEN
              BEGIN
                WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) Underivable tests');
              END;
            CRTable.TestIfNtToTerm(ok)
          END;
        IF ok THEN
          BEGIN
            IF IDE THEN
              BEGIN
                WriteLn(lst); WriteLn(lst, ATGFileName, ' (0, 0) LL(1) tests');
              END;
            CRTable.LL1Test(ll1)
          END;
        WriteLn(lst);
        IF NOT ok OR NOT ll1 OR CRTable.ddt['L'] OR CRTable.ddt['X'] THEN
          BEGIN
            IF NOT IDE THEN Msg('listing'); PrintListing;
            IF CRTable.ddt['X'] THEN CRTable.XRef;
          END;
        IF CRTable.ddt['N'] OR CRTable.symNames THEN
          BEGIN
            IF NOT IDE THEN Msg('symbol name assignment');
            CRTable.AssignSymNames(CRTable.ddt['N'], CRTable.symNames);
          END;
        IF ok AND NOT CRTable.ddt['T'] THEN
          BEGIN
            IF NOT IDE THEN Msg('generating parser');
            CRX.GenCompiler;
            IF CRTable.genScanner AND NOT CRTable.ddt['P'] THEN
              BEGIN
                IF NOT IDE THEN Msg('generating scanner');
                CRA.WriteScanner(ok);
                IF CRTable.ddt['A'] THEN CRA.PrintStates;
              END;
            IF CRTable.ddt['C'] THEN
              BEGIN
                IF NOT IDE THEN Msg('generating compiler');
                CRC.WriteDriver
              END;
            IF NOT IDE THEN CRX.WriteStatistics;
          END;
        IF NOT ok
          THEN
            BEGIN
              IF IDE THEN Write(lst, ATGFileName, ' (0, 0) ');
              Msg('Compilation ended with errors in grammar tests.')
            END
          ELSE IF NOT ll1 THEN
            BEGIN
              IF IDE THEN Write(lst, ATGFileName, ' (0, 0) ');
              Msg('Compilation ended with LL(1) errors.')
            END
          ELSE Msg('Compilation completed. No errors detected.');
      END
    ELSE
      BEGIN
        IF NOT IDE THEN Msg('listing'); PrintListing;
        IF CRTable.ddt['X'] THEN CRTable.XRef;
        IF NOT IDE THEN Msg('*** errors detected ***');
      END;
  WriteLn;
  IF CRTable.ddt['G'] THEN CRTable.PrintGraph;
  IF CRTable.ddt['S'] THEN CRTable.PrintSymbolTable;
  Close(lst);
  Close(src);
END. (* CR *)
